src/v/rpc/connection_cache.h

```diff
@@ -150,40 +156,12 @@ class connection_cache final
     }
 
 private:
+    size_t _max_connections;
     std::optional<connection_cache_label> _label;
     mutex _mutex; // to add/remove nodes
     underlying _cache;
     transport_version _default_transport_version{transport_version::v2};
     ss::gate _gate;
     ss::optimized_optional<ss::abort_source::subscription> _as_subscription;
 };
-inline ss::shard_id connection_cache::shard_for(
-  model::node_id self,
-  ss::shard_id src_shard,
-  model::node_id n,
-  ss::shard_id total_shards) {
-    if (ss::smp::count <= 8) {
-        return src_shard;
-    }
-    static const constexpr size_t vnodes = 8;
-    /// make deterministic - choose 1 prime to mix node_id with
-    /// https://planetmath.org/goodhashtableprimes
-    static const constexpr std::array<size_t, vnodes> universe{
-      {12582917,
-       25165843,
-       50331653,
-       100663319,
-       201326611,
-       402653189,
-       805306457,
-       1610612741}};
-
-    // NOLINTNEXTLINE
-    size_t h = universe[jump_consistent_hash(src_shard, vnodes)];
-    boost::hash_combine(h, std::hash<model::node_id>{}(n));
-    boost::hash_combine(h, std::hash<model::node_id>{}(self));
-    // use self node id to shift jump_consistent_hash_assignment
-    return jump_consistent_hash(h, total_shards);
```

Comment:
I think we need to understand the old logic to see how the new logic will improve things.

Do you understand what the mapping to `vnode` is trying to do?

ISTM it groups all src shards into one of 8 vnodes, and the mapping is the same for all shards with the same vnode. Any idea why that would be desirable?

It definitely seems to have some undesirable properties in that any collisions (or anti-collisions, i.e., vnodes which have 0 shards mapped to them) in the vnode mapping (line 182) will necessarily result in collisions (or anti-collisions) in the final hash value. I.e., applying multiple hashes is generally worse than a single hashes because collisions accumulate.

What I'm confused about is the use of consistent hash (jch) here: jch (vs regular hash) is useful when the second argument ("number of buckets") changes, right? Hashes will change as little as possible when that occurs, that's the "consistent" part? Here, however, `vnodes` is definitely never changing and `total_shards` is I suppose never other than `smp::count` except in tests?

Here's the original change which introduced vnodes:

https://github.com/redpanda-data/redpanda/commit/d3eb4fc6c8ef6716b1524436a34ad64d27e4095c

The basic idea is before there was only 1 connection per broker to each other broker. So this change bumped it by 3. However it does it by mapping each shard to a vnode as described, and I can't extract the reasoning from that commit.